### PR TITLE
Fix bug in torch compile and seqdim is integer

### PR DESCRIPTION
**Pull request by wplf**
# Description

When NVTE_TORCH_COMPILE is 0,  jit_fuser will be torch.jit.script.
And the compiled function don't accept interger. Then, bug will occur in the function below  due to the `seq_dim` argument.

```
@jit_fuser
def flash_attn_fwd_out_correction(out, out_per_step, seq_dim, softmax_lse, softmax_lse_per_step):
    """Merge partial outputs of each step in Attention with context parallelism"""
    softmax_lse_corrected_exp = torch.exp(softmax_lse_per_step - softmax_lse).movedim(2, seq_dim)
    softmax_lse_corrected_exp = softmax_lse_corrected_exp.unsqueeze(-1)
    out_corrected = out_per_step * softmax_lse_corrected_exp
    out.add_(out_corrected)
```

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

Please list the changes introduced in this PR:

- Change A
- Change B

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by ksivaman — CHANGES_REQUESTED** [review]
@wplf Can we not annotate `flash_attn_fwd_out_correction` instead? IIUC, the fusers assume every argument is a tensor unless explicitly specified.

**Comment by wplf** [comment]
> @wplf Can we not annotate `flash_attn_fwd_out_correction` instead? IIUC, the fusers assume every argument is a tensor unless explicitly specified.

That's right. Torch.jit.fuser need arguments to be tensor, but torch.compiler don't. 
I am afraid removing the annotation will degrade performance in the case of `torch.complie`.

**Comment by ksivaman** [comment]
Let's try and annotate the function and see if it fixes the error!

**Comment by wplf** [comment]
> Let's try and annotate the function and see if it fixes the error!

I have checked that removing the annotation will fix the problem. And the pull request is changed, too.

**Comment by ksivaman** [comment]
I don't see the changes yet, could you update the PR?

**Comment by wplf** [comment]
> I don't see the changes yet, could you update the PR?

Sorry, I don't expect you so quickly like that. You can check it now.

File: transformer_engine/pytorch/attention.py [hunk 1]
```diff
@@ -1363,7 +1363,7 @@ def flash_attn_p2p_communicate(
     return send_recv_reqs
 
 
-@jit_fuser
+@torch.compile
 def flash_attn_fwd_out_correction(out, out_per_step, seq_dim, softmax_lse, softmax_lse_per_step):
```
Comment (ksivaman) [inline transformer_engine/pytorch/attention.py:1367]:
```suggestion
@jit_fuser
def flash_attn_fwd_out_correction(
    out: torch.Tensor,
    out_per_step: torch.Tensor,
    seq_dim: int,
    softmax_lse: ... #TODO,
    softmax_lse_per_step: ...#TODO,
):
```

**Comment by ksivaman** [comment]
We still want the `jit_fuser` decorator instead of forcing it to be `torch.compile`. What I meant by annotating is to add explicit types to the arguments. See given suggestion (just an example)

**Comment by ksivaman** [comment]
@wplf Could you see if the new changes resolve the error?

**Comment by xrennvidia** [comment]
LGTM.

**Pull request merged**